File: apps/mobile/lib/features/access_payment/data/dtos/access_session_dto.dart
import "../../domain/entities/access_session.dart";
import "../../domain/entities/access_session_status.dart";

/// JSON mapping for the `AccessSession` schema in docs/api/openapi.yaml —
/// shared by `POST /access-sessions` (201) and `GET /access-sessions/{id}`
/// (200). Field set matches the schema exactly; see
/// `AccessSession`'s own doc comment for why `qrCodeScanned`, `userId`,
/// and `transaction` are deliberately absent.
class AccessSessionDto {
  const AccessSessionDto({
    required this.id,
    required this.cabinId,
    required this.status,
    required this.startedAt,
    required this.unlockedAt,
  });

  factory AccessSessionDto.fromJson(Map<String, dynamic> json) {
    return AccessSessionDto(
      id: json["id"] as String,
      cabinId: json["cabinId"] as String,
      status: json["status"] as String,
      startedAt: json["startedAt"] as String,
      unlockedAt: json["unlockedAt"] as String?,
    );
  }

  final String id;
  final String cabinId;

  /// Wire value is `snake_case` (e.g. `payment_pending`), not the Dart
  /// enum's `camelCase` name — mapped explicitly below, same discipline
  /// as `StationDetailDto.configuration`.
  final String status;
  final String startedAt;
  final String? unlockedAt;

  AccessSession toEntity() {
    return AccessSession(
      id: id,
      cabinId: cabinId,
      status: switch (status) {
        "initiated" => AccessSessionStatus.initiated,
        "payment_pending" => AccessSessionStatus.paymentPending,
        "unlocked" => AccessSessionStatus.unlocked,
        "in_use" => AccessSessionStatus.inUse,
        "completed" => AccessSessionStatus.completed,
        "cancelled" => AccessSessionStatus.cancelled,
        _ => AccessSessionStatus.initiated,
      },
      startedAt: DateTime.parse(startedAt),
      unlockedAt: unlockedAt == null ? null : DateTime.parse(unlockedAt!),
    );
  }
}

File: apps/backend/src/modules/access-payment/interface/dto/access-session.dto.ts
import { ApiProperty } from '@nestjs/swagger';
import { IsString, MaxLength, MinLength } from 'class-validator';
import { AccessSession } from '../../domain/entities/access-session.entity';

const ACCESS_SESSION_STATUSES = ['initiated', 'payment_pending', 'unlocked', 'in_use', 'completed', 'cancelled'] as const;

/** Matches openapi.yaml components.schemas.AccessSessionCreateRequest exactly. */
export class AccessSessionCreateRequestDto {
  @ApiProperty()
  @IsString()
  @MinLength(1)
  @MaxLength(500)
  qrCodeScanned!: string;
}

/** Matches openapi.yaml components.schemas.AccessSession exactly. */
export class AccessSessionResponseDto {
  @ApiProperty({ format: 'uuid' })
  id!: string;

  @ApiProperty({ format: 'uuid' })
  cabinId!: string;

  @ApiProperty({ enum: ACCESS_SESSION_STATUSES })
  status!: string;

  @ApiProperty({ format: 'date-time' })
  startedAt!: string;

  @ApiProperty({ format: 'date-time', nullable: true })
  unlockedAt!: string | null;

  static fromDomain(session: AccessSession): AccessSessionResponseDto {
    const dto = new AccessSessionResponseDto();
    dto.id = session.id;
    dto.cabinId = session.cabinId;
    dto.status = session.status;
    dto.startedAt = session.startedAt.toISOString();
    dto.unlockedAt = session.unlockedAt ? session.unlockedAt.toISOString() : null;
    return dto;
  }
}
